Fabric.Authorization.API/scripts/AzureADUsers-Migration.ps1

```diff
@@ -154,130 +220,158 @@ function Add-AuthUsers
 			  }
 			  $command.Parameters.RemoveAt("@parentId")
 
-			  if(![string]::IsNullOrEmpty($user.GroupId))
-			  {
-				$command.CommandText = "IF NOT EXISTS 
-										(SELECT 1 FROM [dbo].[GroupUsers] g
-										 WHERE SubjectId = @subjectId
-										 AND IdentityProvider = 'AzureActiveDirectory')
-										BEGIN
-										INSERT INTO [dbo].[GroupUsers] ([SubjectId], [IdentityProvider], [CreatedBy], [CreatedDateTimeUtc], [GroupId], [IsDeleted])
-										VALUES(@subjectId, @identityProvider, @createdBy, @createdDateTime, @groupId, @isDeleted)
-										END;"
-				$command.Parameters["@subjectId"].Value = $user.objectId
-				$command.Parameters["@identityProvider"].Value = "AzureActiveDirectory"
-				$command.Parameters["@createdBy"].Value = "AzureADUsers-Migration"
-				$command.Parameters["@createdDateTime"].Value = $sqlDate
-				$command.Parameters.Add("@groupId", [System.Data.SqlDbType]::UniqueIdentifier)
-				$command.Parameters["@groupId"].Value = $user.GroupId
-				$command.Parameters["@isDeleted"].Value = 0
-				$groupCount++
+            $groupTable = $authDataSet.Tables[1]
+		    foreach ($group in $groupTable)
+		    {
+			    if(![string]::IsNullOrEmpty($group.GroupId) -and $user.Id -eq $group.Id)
+			    {
+			    $command.CommandText = "IF NOT EXISTS 
+									    (SELECT 1 FROM [dbo].[GroupUsers] g
+										    WHERE SubjectId = @subjectId
+                                            AND GroupId = @groupId
+										    AND IdentityProvider = 'AzureActiveDirectory')
+									    BEGIN
+									    INSERT INTO [dbo].[GroupUsers] ([SubjectId], [IdentityProvider], [CreatedBy], [CreatedDateTimeUtc], [GroupId], [IsDeleted])
+									    VALUES(@subjectId, @identityProvider, @createdBy, @createdDateTime, @groupId, @isDeleted)
+									    END;"
+			    $command.Parameters["@subjectId"].Value = $user.objectId
+			    $command.Parameters["@identityProvider"].Value = "AzureActiveDirectory"
+			    $command.Parameters["@createdBy"].Value = "AzureADUsers-Migration"
+			    $command.Parameters["@createdDateTime"].Value = $sqlDate
+			    $command.Parameters.Add("@groupId", [System.Data.SqlDbType]::UniqueIdentifier)
+			    $command.Parameters["@groupId"].Value = $group.GroupId
+			    $command.Parameters["@isDeleted"].Value = 0
+			    $groupCount++
 
                 $resultGroupUsers = $command.ExecuteNonQuery()
                 if($resultGroupUsers -gt 0)
                 {   
-			      $totalGroupUsersAdded += $resultGroupUsers
+			        $totalGroupUsersAdded += $resultGroupUsers
 	    	    }
 
 			    if($groupCount -eq 1)
 			    {
-			      $command.Parameters.RemoveAt("@groupId")
-				  $groupCount = 0
+			        $command.Parameters.RemoveAt("@groupId")
```

Comment:
It's tough to tell from the formatting, but is this inside the foreach loop? Are we using it as a way to clear out that parameter on each iteration of the loop so we can update the groupId param for each iteration?

Reply:
I actually could move that out of the foreach loop, before the next foreach loop for RoleUsers. Parameterized queries don't like unused parameters, and an error is thrown if there are any unused. Since GroupUsers and RoleUsers have at least one different column, I have to remove it before executing the next query.